File: Java/app/src/main/java/com/github/shin1x1/jsonparser/lexer/Lexer.java
package com.github.shin1x1.jsonparser.lexer;

import com.github.shin1x1.jsonparser.lexer.exception.UnexpectedCharacterException;
import com.github.shin1x1.jsonparser.lexer.exception.UnexpectedEotException;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.Optional;

public final class Lexer {
    @Nonnull
    private final Scanner scanner;

    public Lexer(@Nonnull Scanner scanner) {
        this.scanner = scanner;
    }

    public Optional<Token> getNextToken() throws IOException {
        var ch = scanner.consume();

        if (ch.isEmpty()) {
            return Optional.empty();
        }

        char c = ch.get();
        if (c == ' ' || c == '\n' || c == '\r' || c == '\t') {
            return getNextToken();
        }

        return Optional.of(lex(c));
    }

    private Token lex(char ch) throws IOException {
        return switch (ch) {
            case '[' -> new Token.LeftBracket();
            case ']' -> new Token.RightBracket();
            case '{' -> new Token.LeftBrace();
            case '}' -> new Token.RightBrace();
            case ':' -> new Token.Colon();
            case ',' -> new Token.Comma();
            case '"' -> lexString();
            case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' -> lexNumber(ch);
            case 't' -> lexLiteral("true", new Token.True());
            case 'f' -> lexLiteral("false", new Token.False());
            case 'n' -> lexLiteral("null", new Token.Null());
            default -> throw new UnexpectedCharacterException(ch);
        };
    }

    private Token lexNumber(char first) throws IOException {
        var chs = new StringBuilder(String.valueOf(first));

        while (true) {
            var och = scanner.peek();
            if (och.isEmpty()) {
                break;
            }

            var ch = och.get();
            if (!"0123456789+-e.".contains(ch.toString())) {
                break;
            }

            scanner.consume();
            chs.append(ch);
        }

        return new Token.Number(chs.toString());
    }

    private Token lexString() throws IOException {
        var chs = new StringBuilder();
        var backslash = false;

        while (true) {
            var ch = scanner.consume().orElseThrow(UnexpectedEotException::new);

            if (!backslash) {
                if (ch == '"') {
                    return new Token.String(chs.toString());
                }

                if (ch == '\\') {
                    backslash = true;
                    continue;
                }

                chs.append(ch);
                continue;
            }

            backslash = false;
            chs.append(
                    switch (ch) {
                        case '"' -> '"';
                        case '\\' -> '\\';
                        case '/' -> '/';
                        case 'b' -> '\b';
                        case 'f' -> '\f';
                        case 'n' -> '\n';
                        case 'r' -> '\r';
                        case 't' -> '\t';
                        case 'u' -> lexCodepoint();
                        default -> throw new UnexpectedCharacterException(ch);
                    }
            );
        }
    }

    private char lexCodepoint() throws IOException {
        var code = 0;
        for (var i = 0; i < 4; i++) {
            var ch = scanner.consume().orElseThrow(UnexpectedEotException::new);

            if (!"0123456789ABCDEF".contains(ch.toString())) {
                throw new UnexpectedCharacterException(ch);
            }

            code += (ch - '0') * Math.pow(16, 3 - i);
        }

        return (char) code;
    }

    private Token lexLiteral(String literal, Token token) throws IOException {
        for (int i = 1; i < literal.length(); i++) {
            var ch = scanner.consume().orElseThrow(UnexpectedEotException::new);
            if (ch != literal.charAt(i)) {
                throw new UnexpectedCharacterException(ch);
            }
        }

        return token;
    }
}

File: Scala3/src/main/scala/lexer/Lexer.scala
package lexer

import scala.annotation.tailrec
import scala.math.pow
import scala.util.{Failure, Success, Try}

enum Token:
  case LeftBracket, RightBracket
  case LeftBrace, RightBrace
  case Colon, Comma
  case Null, True, False
  case Number(value: Double)
  case String(value: scala.Predef.String)

class InvalidCharacter(val char: Char) extends Exception

class InvalidCodepoint() extends Exception

class InvalidLiteralCharacter() extends Exception

type TryToken = Try[Token]

object Lexer:
  def getNextToken(scanner: Scanner): TryToken = scanner.consume() match
    case Some(c) => lex(c)(using scanner)
    case None => Failure(IndexOutOfBoundsException())

  private def lex(char: Char)(using Scanner): TryToken = char match
    case '[' => Success(Token.LeftBracket)
    case ']' => Success(Token.RightBracket)
    case '{' => Success(Token.LeftBrace)
    case '}' => Success(Token.RightBrace)
    case ':' => Success(Token.Colon)
    case ',' => Success(Token.Comma)
    case '-' | '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => lexNumber(char)
    case '"' => lexString(char)
    case 't' => lexLiteral("true", Token.True)
    case 'f' => lexLiteral("false", Token.False)
    case 'n' => lexLiteral("null", Token.Null)
    case x => Failure(InvalidCharacter(x))

  private def lexNumber(char: Char)(using scanner: Scanner): TryToken =
    @tailrec
    def lex(string: String): String = scanner.peek() match
      case Some(c) if "-+.e0123456789".contains(c) => {
        scanner.consume()
        lex(string :+ c)
      }
      case _ => string

    try
      Success(Token.Number(lex(char.toString).toDouble))
    catch
      case e: Exception => Failure(e)

  private def lexString(char: Char)(using scanner: Scanner): TryToken =
    def lexCodepoint(): Option[Char] =
      val isHex = (c: Char) => "0123456789ABCDEF".contains(c)
      for
        h1 <- scanner.consume().filter(isHex)
        h2 <- scanner.consume().filter(isHex)
        h3 <- scanner.consume().filter(isHex)
        h4 <- scanner.consume().filter(isHex)
      yield
        ((h1 - '0' << 12) + (h2 - '0' << 8) + (h3 - '0' << 4) + h4 - '0').toChar

    @tailrec
    def lex(string: String, backslash: Boolean = false): TryToken = scanner.consume() match
      case Some(s) if backslash => {
        val ch = s match
          case '"' => '"'
          case '\\' => '\\'
          case '/' => '/'
          case 'b' => '\b'
          case 'f' => '\f'
          case 'n' => '\n'
          case 'r' => '\r'
          case 't' => '\t'
          case 'u' => lexCodepoint() match
            case Some(c1) => c1
            case None => return Failure(InvalidCodepoint())
          case _ => '?'

        lex(string :+ ch, false)
      }
      case Some(s) if s == '"' => Success(Token.String(string))
      case Some(s) if s == '\\' => lex(string, true)
      case Some(s) => lex(string :+ s)
      case None => Failure(Exception())

    lex("")

  private def lexLiteral(literal: String, token: Token)(using scanner: Scanner): TryToken =
    // skip head character in literal
    literal.substring(1).foreach(c => scanner.consume() match
      case Some(s) if s == c => ()
      case _ => return Failure(InvalidLiteralCharacter())
    )

    Success(token)


